private_set_intersection/go/client/client.go

```diff
@@ -110,36 +148,115 @@ func (c *PsiClient) CreateRequest(rawInput []string) (string, error) {
 	var err *C.char
 
 	rcode := C.psi_client_create_request(c.context, &inputs[0], C.size_t(len(inputs)), &out, &outlen, &err)
+
+	for idx := range inputs {
+		C.free(unsafe.Pointer(inputs[idx].buff))
+	}
+
 	if rcode != 0 {
 		return "", fmt.Errorf("create request failed %v(%v)", c.loadCString(&err), rcode)
 	}
 
 	return c.loadCString(&out), nil
 }
 
-//ProcessResponse processes the server's response and returns the PSI cardinality. The
-//first argument, `server_setup`, is a bloom filter that encodes encrypted
-//server elements and is sent by the server in a setup phase. The second
-//argument, `server_response`, is the response received from the server
-//after sending the result of `CreateRequest`.
+//GetIntersection processes the server's response and returns the intersection of the client
+//and server inputs. Use this function if this instance was created with
+//`reveal_intersection = true`. The first argument, `server_setup`, is a
+//bloom filter that encodes encrypted server elements and is sent by the
+//server in a setup phase. The second argument, `server_response`, is the
+//response received from the server after sending the result of
+//`CreateRequest`.
+//
+//Returns INVALID_ARGUMENT if any input messages are malformed, or INTERNAL
+//if decryption fails.
+func (c *PsiClient) GetIntersection(serverSetup, serverResponse string) ([]int64, error) {
+	if c.context == nil {
+		return nil, errors.New("invalid context")
+	}
+
+	var out *C.int64_t
+	var outlen C.size_t
+	var err *C.char
+
+	csetup := C.CString(serverSetup)
+	defer C.free(unsafe.Pointer(csetup))
+
+	cresponse := C.CString(serverResponse)
+	defer C.free(unsafe.Pointer(cresponse))
+
+	rcode := C.psi_client_get_intersection(c.context, csetup, cresponse, &out, &outlen, &err)
+
+	if rcode != 0 {
+		return nil, fmt.Errorf("process response failed: %v(%v)", c.loadCString(&err), rcode)
+	}
+
+	resultPtr := unsafe.Pointer(out)
+	resultLen := int64(outlen)
+
+	var result []int64
+	var idx int64
+	var ref C.int64_t
+
+	for idx = 0; idx < resultLen; idx++ {
+		val := (*int64)(unsafe.Pointer(uintptr(resultPtr) + uintptr(idx)*unsafe.Sizeof(&ref)))
```

Comment:
This seems kind of unsafe... Doesn't `unsafe.Sizeof(&ref)` return the size of a int64 pointer, not the size of an int64?

Reply:
I tried to improve it in f33eb05ca1be811f46764f973d36c2464b5130f3, please have a look.